src/smt/theory_str_noodler/decision_procedure.cpp

```diff
@@ -1166,22 +1173,101 @@ namespace smt::noodler {
+        // a stable solution
+        for(const Predicate& dis : this->disequations.get_predicates()) {
+            proj_diseqs.add_predicate(Predicate(PredicateType::Inequation, {
+                proj_concat(dis.get_left_side()),
+                proj_concat(dis.get_right_side()),
+            }));
+        }
+
+        STRACE("str", tout << "CA-DISEQS (original): " << std::endl << this->disequations.to_string() << std::endl;);
+        STRACE("str", tout << "CA-DISEQS (substituted): " << std::endl << proj_diseqs.to_string() << std::endl;);
+        return ca::get_lia_for_disequations(proj_diseqs, this->solution.aut_ass);
+    }
+
+    std::pair<LenNode, LenNodePrecision> DecisionProcedure::get_formula_for_not_contains() {
+        Formula proj_not_cont {};
+
+        auto proj_concat = [&](const Concat& con) -> Concat {
+            Concat ret {};
+            for(const BasicTerm& bt : con) {
+                Concat subst = this->solution.get_substituted_vars(bt);
+                ret.insert(ret.end(), subst.begin(), subst.end());
+            }
+            return ret;
+        };
+
+        // take the original disequations (taken from input) and 
+        // propagate substitutions involved by the current substitution map of 
+        // a stable solution
+        for(const Predicate& dis : this->not_contains.get_predicates()) {
+            proj_not_cont.add_predicate(Predicate(PredicateType::NotContains, {
+                proj_concat(dis.get_left_side()),
+                proj_concat(dis.get_right_side()),
+            }));
+        }
+
+        STRACE("str", tout << "CA-DISEQS (original): " << std::endl << this->not_contains.to_string() << std::endl;);
+        STRACE("str", tout << "CA-DISEQS (substituted): " << std::endl << proj_not_cont.to_string() << std::endl;);
+        return ca::get_lia_for_not_contains(proj_not_cont, this->solution.aut_ass);
+    }
+
     /**
      * @brief Creates initial inclusion graph according to the preprocessed instance.
      */
     void DecisionProcedure::init_computation() {
         Formula equations;
+
+        bool single_diseq = false;
+
         for (auto const &dis_or_eq : formula.get_predicates()) {
             if (dis_or_eq.is_equation()) {
                 equations.add_predicate(dis_or_eq);
             } else if (dis_or_eq.is_inequation()) {
-                for (auto const &eq_from_diseq : replace_disequality(dis_or_eq)) {
-                    equations.add_predicate(eq_from_diseq);
+                // if we solve diesquations using CA --> we store the disequations to be solved later on
+                if(!single_diseq && this->m_params.m_ca_constr) {
+                    init_ca_diseq(dis_or_eq);
+                    single_diseq = true;
+                } else {
+                    for (auto const &eq_from_diseq : replace_disequality(dis_or_eq)) {
+                        equations.add_predicate(eq_from_diseq);
+                    }
```

Comment:
How does this works? If we have let's say 5 disequations, one of them is handled trough CA, and others are handled by the way we did it before (trough `to_code`)?

Reply:
Yes, Michal works on extension to arbitrary number of disequations.

Reply:
Ok, I would just call the bool variable differently, because right now I thought it should be checking if we have one disequation. Maybe `some_diseq_handled_by_ca`?

Reply:
Nice, fixed.